Framework/HistogramData/inc/MantidHistogramData/Histogram.h

```diff
@@ -0,0 +1,462 @@
+  Histogram &operator=(Histogram &&)& = default;
+
+  XMode xMode() const noexcept { return m_xMode; }
+
+  BinEdges binEdges() const;
+  BinEdgeVariances binEdgeVariances() const;
+  BinEdgeStandardDeviations binEdgeStandardDeviations() const;
+  Points points() const;
+  PointVariances pointVariances() const;
+  PointStandardDeviations pointStandardDeviations() const;
+  template <typename... T> void setBinEdges(T &&... data) & ;
+  template <typename... T> void setBinEdgeVariances(T &&... data) & ;
+  template <typename... T> void setBinEdgeStandardDeviations(T &&... data) & ;
+  template <typename... T> void setPoints(T &&... data) & ;
+  template <typename... T> void setPointVariances(T &&... data) & ;
+  template <typename... T> void setPointStandardDeviations(T &&... data) & ;
+
+  Counts counts() const;
+  CountVariances countVariances() const;
+  CountStandardDeviations countStandardDeviations() const;
+  Frequencies frequencies() const;
+  FrequencyVariances frequencyVariances() const;
+  FrequencyStandardDeviations frequencyStandardDeviations() const;
+  template <typename... T> void setCounts(T &&... data) & ;
+  template <typename... T> void setCountVariances(T &&... data) & ;
+  template <typename... T> void setCountStandardDeviations(T &&... data) & ;
+  template <typename... T> void setFrequencies(T &&... data) & ;
+  template <typename... T> void setFrequencyVariances(T &&... data) & ;
+  template <typename... T> void setFrequencyStandardDeviations(T &&... data) & ;
+
+  const HistogramX &x() const { return *m_x; }
+  const HistogramY &y() const { return *m_y; }
+  const HistogramE &e() const { return *m_e; }
+  const HistogramDx &dx() const { return *m_dx; }
+  HistogramX &mutableX() & { return m_x.access(); }
+  HistogramY &mutableY() & { return m_y.access(); }
+  HistogramE &mutableE() & { return m_e.access(); }
+  HistogramDx &mutableDx() & { return m_dx.access(); }
+
+  Kernel::cow_ptr<HistogramX> sharedX() const { return m_x; }
+  Kernel::cow_ptr<HistogramY> sharedY() const { return m_y; }
+  Kernel::cow_ptr<HistogramE> sharedE() const { return m_e; }
+  Kernel::cow_ptr<HistogramDx> sharedDx() const { return m_dx; }
+  void setSharedX(const Kernel::cow_ptr<HistogramX> &x) & ;
+  void setSharedY(const Kernel::cow_ptr<HistogramY> &y) & ;
+  void setSharedE(const Kernel::cow_ptr<HistogramE> &e) & ;
+  void setSharedDx(const Kernel::cow_ptr<HistogramDx> &Dx) & ;
+
+  // Temporary legacy interface to X
+  void setX(const Kernel::cow_ptr<HistogramX> &X) & { m_x = X; }
+  MantidVec &dataX() & { return m_x.access().mutableRawData(); }
+  const MantidVec &dataX() const & { return m_x->rawData(); }
+  const MantidVec &readX() const { return m_x->rawData(); }
+  Kernel::cow_ptr<HistogramX> ptrX() const { return m_x; }
+
+  // Temporary legacy interface to Y
+  void setY(const Kernel::cow_ptr<HistogramY> &Y) & { m_y = Y; }
+  MantidVec &dataY() & { return m_y.access().mutableRawData(); }
```

Comment:
I think we cannot do anything about this as long as the old interface is around. Client code using the old interface simply requires length modifications, so without refactoring to use the new interface we are stuck with this.
If the vector is extended nothing bad happens to `FixedLengthVector`. The only risk is that the `Histogram` can get into an inconsistent state where `X` and `Y` length do not match, but this is the same issue we have right now. By using the new interface in more and more places we are reducing this risk step by step.